virtcontainers/sandbox.go

```diff
@@ -940,7 +940,13 @@ func (s *Sandbox) startVM() error {
 	// Once startVM is done, we want to guarantee
 	// that the sandbox is manageable. For that we need
 	// to start the sandbox inside the VM.
-	return s.agent.startSandbox(s)
+	err := s.agent.startSandbox(s)
+	defer func() {
```

Comment:
is `defer` needed? why not just 

```go
if err != nil {
	s.hypervisor.stopSandbox()
}
```
:thinking:

Reply:
or

```go
if err != nil {
	return s.hypervisor.stopSandbox()
}
```

in case of  `stopSandbox` fails

Reply:
Thanks! I agree with your last proposed code in case of `stopSandbox` fails

Reply:
This is not semantically correct. We should not call `stopSandbox()` after `startSandbox()` returned some error. We would be running `stopSandbox()`, which assumes the sandbox has been properly started, when this is not true.

The valid case for such a thing could be to run `stopSandbox()` if further in the code, an error would occur, and we would want to rollback from there.

And if you want to do such thing, I think it should be part of its own commit for more clarity ;)

Reply:
Maybe we have some misunderstanding about `s.agent.stopSandbox`() and `s.hypervisor.stopSandbox()`.

`s.agent.stopSandbox()` is used to stop the Sandbox and Containers which are running in VM, in other words, both sandbox and containers are started successfully in VM.

`s.hypervisor.stopSandbox()` is used to stop qemu process and release related resources, in other words, sandbox is not created successfully in VM.

In this place, if `s.agent.startSandbox()` return error , it indicates that sandbox is not created successfully in VM, so I want to call `s.hypervisor.stopSandbox()` to stop VM process.